### ISSUE TITLE
Reply to Netbout message from Email

### ISSUE THREAD

**Issue by dmzaytsev** [yegor256/netbout#600]
At the moment when I get a email notification to respond to the message, I have to open a browser, to login to Netbout and type a message.
It would be useful to be able to reply  from email as it does Github.

**Comment by yegor256** [comment]
@dmzaytsev yes, make sense, thanks

**Comment by dmzaytsev** [comment]
set milestone to 3.0

**Comment by karato** [comment]
@dmzaytsev I am aware of the task, give me some time to find a developer...

**Comment by karato** [comment]
@dmzaytsev since the ticket has no milestone I set it to 3.1

**Comment by karato** [comment]
@dmzaytsev thanks for tis bug, I topped your account for 15 mins, transaction `57638899`

**Comment by karato** [comment]
> @karato set milestone to 3.0

@dmzaytsev milestone set to `3.0`, thanks

**Comment by karato** [comment]
@ikhvostenkov this is your task

**Comment by dmzaytsev** [comment]
@ikhvostenkov let's use [GreenMail](http://www.icegreen.com/greenmail/) for email testing

**Comment by ikhvostenkov** [comment]
@dmzaytsev ```EmMessages```:
```Java
private void email(final Friend friend, final String text)
        throws IOException {
        this.postman.send(
            new Envelope.MIME()
                .with(new StRecipient(friend.alias(), friend.email()))
                .with(
                    new StSubject(
                        String.format(
                            "#%d: %s",
                            this.bout.number(),
                            this.bout.title()
                        )
                    )
                )
                .with(
                    new EnHTML(
                        Joiner.on('\n').join(
                            new Markdown(text).html(),
                            "<p>--<br/>to reply click here: ",
                            String.format(
                                "http://www.netbout.com/b/%d</p>",
                                this.bout.number()
                            ),
                            "</p>"
                        )
                    )
                )
        );
    }
```
This is already implemented. Or do you mean some other functionality?

**Comment by dmzaytsev** [comment]
@ikhvostenkov I mean reply into a bout from the email response.

* for security reason and avoid spam we have to generate an unique string for  the message id, when we sending email we have to write to a table  message_id, bout_id, author_alias

* We have to add header `Reply-To: <message_id>@reply.netbout.com>` into an outgoing email. e.g `Reply-To: jhjkh7y78y@reply.netbout.com>` 
* netbout  should monitor mailbox `bouts@reply.netbout.com` 
* nebout recieve email with the header `To:<message_id>@reply.netbout.com`
* looking for boat_id, author_alias,
* post the message as the author
* delete message_id from table.


btw @yegor256 can we setup a redirection all incoming messages for `reply.netbout.com` into `bouts@reply.netbout.com` ? 
@ikhvostenkov  anyway you shouldn't use a hardcoded  box names.

@yegor256 are you agree with the  proposed design?

**Comment by yegor256** [comment]
@dmzaytsev FYI, we have this feature in the previous version of Netbout (1.6), see this code: https://github.com/yegor256/netbout/blob/netbout-1.6/netbout-rest/src/main/java/com/netbout/notifiers/email/RoutineFarm.java So, yes, I agree with you :)

**Comment by dmzaytsev** [comment]
@yegor256 why you removed this feature?

**Comment by yegor256** [comment]
@dmzaytsev I was too lazy to implement it. Version 2.0 of Netbout was created mostly from scratch, only some elements from 1.6 were used. I didn't take this element since I didn't have time.

**Comment by dmzaytsev** [comment]
@ikhvostenkov any updates here?
You do not need to implement the entire task, it is too large for a single ticket. Let's use a puzzle www.yegor256.com/2009/03/04/pdd.html.
Feel free to ask any technical questions here.

**Comment by ikhvostenkov** [comment]
@dmzaytsev Could you, please, assign someone else or postpone this task until 8th of June, I will be out until this for personal reasons?

**Comment by dmzaytsev** [comment]
@yegor256 let's make this task simpler
* we have to setup a redirection all incoming messages for `@reply.netbout.com` into `bouts@reply.netbout.com` 
* we have to add header `Reply-To: <boat_id>@reply.netbout.com>` into an outgoing email. e.g `Reply-To: 35351@reply.netbout.com>`
* netbout should monitor mailbox bouts@reply.netbout.com
* nebout recieve email with the header To:<boat_id>@reply.netbout.com
looking for the alias by email and boat_id
* post the message as the author

Are you agree?

**Comment by dmzaytsev** [comment]
@yegor256 please ask `@karato` assign this task to someone else

**Comment by yegor256** [comment]
@dmzaytsev yes, looks correct to me

**Comment by yegor256** [comment]
@karato assign someone else here pls

**Comment by dmzaytsev** [comment]
@karato this is urgent

**Comment by karato** [comment]
> @karato assign someone else here pls

@yegor256 -30 points to your rating

**Comment by karato** [comment]
> @karato assign someone else here pls

@yegor256 someone else will help in this task, no problem at all

**Comment by karato** [comment]
> @karato this is urgent

@dmzaytsev thanks, I added the "urgent" tag

**Comment by karato** [comment]
@erimerturk this is your task

**Comment by erimerturk** [comment]
@dmzaytsev you have already defined steps but also could we decide which steps will be implemented in this bug ? which of them should be puzzle ? what is your suggestion ? thank you

**Comment by dmzaytsev** [comment]
@erimerturk 
let's implement the mailbox monitor in this task.
something like this
```
class EmCatch {
    EmCatch (Postman postman, Action action)
    interface Action{
        run(Boat boat, String email, String Text);
    }
}
```
Monitor shoud use single daemon thread inside.
and don't forget about the unit tests. You might to use [GreenMail](http://www.icegreen.com/greenmail/) for the unit tests and `Ports` for the random port reservation.
Other steps must be described in the puzzle

**Comment by erimerturk** [comment]
@dmzaytsev your description is not enough i think. Could you please explain <b>EmCatch</b> object more ? What is the duty of <b>EmCatch</b> , how will you initialize it, which interface do you need ? I need more information about <b>EmCatch</b>. This version of it is not clear for me, thank you.

**Comment by dmzaytsev** [comment]
@erimerturk Sorry I thought jcabi-email supports not sending only...
well.. the `EmCatch` mission is periodically(in separate thread) grab all messages from mailbox(POP3 or IMAP) and to call `EmCatch.Action.run()` for each message. We should pass into the ctor the connection parameters (javax.mail API) and the `EmCacth.Action` instance. We'll create `EmCatch` in `com.netbout.Launch`.

You don't need to implement `EmCacth.Action` in this task, another developer will do it. Just implement `EmCatch`.

**Comment by dmzaytsev** [comment]
@erimerturk any progress here?

**Comment by erimerturk** [comment]
@dmzaytsev i am working on it but i should say that it is a little bit difficult for me. could you please review my first version from [gist](https://gist.github.com/erimerturk/f0832f1d8a874f9b82af).  i think your comment could help me.

**Comment by dmzaytsev** [comment]
@erimerturk I would do it like this
```
final class EmCatch {
    /**
     * Action.
     */
    private final transient Action action;
    /**
     * Email Properties.
     */
    private final transient Properties properties;

    EmCatch (final Action act, final Properties pro){
        this.action = act;
        this.properties = pro;
        final Thread monitor = new Thread(
            new Runnable() {
                @Override
                public void run() {
                    while (true) {
                        EmCatch.this.check();
                        try {
                            TimeUnit.SECONDS.sleep( Long.valueOf(
                                Manifests.read("Netbout-MailCheckPeriod")
                            ));
                        } catch (final InterruptedException ex) {
                            Thread.currentThread().interrupt();
                            throw new IllegalStateException(ex);
                        }
                    }
                }
            }
        );
        monitor.setDaemon(true);
        monitor.start();
    }

    private void check() {
            Session session = Session.getDefaultInstance(this.properties);
            try {
                Store store = session.getStore("pop3");
                store.connect(
                    this.properties.getProperty("host"),
                    this.properties.getProperty("user"),
                    this.properties.getProperty("password")
                );
                Folder inbox = store.getFolder("inbox");
                inbox.open(Folder.READ_WRITE);
                FlagTerm unseenFlagTerm = new FlagTerm(new Flags(Flags.Flag.SEEN), false);
                Message messages[] = inbox.search(unseenFlagTerm);
//it would better to use foreach here                 
                for (int i = 0; i < messages.length; i++) {
                    // looking for the boat by email and pass it into action .run()
                    action.run(
                        boat,  
                        messages[i].getFrom()[0], 
                    //  add message text here
                    );
                    
                }
                inbox.close(true);
                store.close();

            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    /**
     * Email Catch Action.
     * @author Erim Erturk (erimerturk@gmail.com)
     * @version $Id$
     * @since 2.14
     */
    interface Action {
        void run(Boat boat, String email, String Text);
    }

}

```

**Comment by dmzaytsev** [comment]
@yegor256 pop3 will be enough for the mailbox monitor or we have to add imap as an option? what do you think?

**Comment by dmzaytsev** [comment]
@erimerturk also please put check period into properties
I think it would better to use `Map` instead `Properties`

**Comment by erimerturk** [comment]
@dmzaytsev  <b>// looking for the boat by email and pass it into action .run()</b>  how can i get email ? Is it equal to <code>messages[i].getFrom()[0] </code> ? 
Is it user email or bout_id ? I need user_email and bout_id both of them. But i think we are losing one of them.

**Comment by dmzaytsev** [comment]
@erimerturk sorry, I meant look for the boat by boat_id
the user replies to the `<boat_id>@reply.netbout.com>` so we should extract `boat_id` from `message.getTo`. The alias email we might take from `message.getFrom`

**Comment by erimerturk** [comment]
@dmzaytsev please review last version from [gist] (https://gist.github.com/erimerturk/8a58969cbac35261711e) . <code> Action.run </code> is expecting Message for now, it should be solved by another issue i think. What do you think ? If  [gist] (https://gist.github.com/erimerturk/8a58969cbac35261711e) version is okay for you, i will send PR. thank you so much.

**Comment by yegor256** [comment]
@dmzaytsev I think POP3 is more than enough. Don't see a reason for using IMAP

**Comment by dmzaytsev** [comment]
@erimerturk let's use the ctor params for user,passwd,host,port,check_period instead the Map
we should have default values for port(110 for pop3) and check_periiod (2min)
then make the PR 
> and don't forget about the unit tests. You might to use GreenMail for the unit tests and Ports for the random port reservation.

**Comment by erimerturk** [comment]
@dmzaytsev please see PR #679

**Comment by dmzaytsev** [comment]
@erimerturk thanks!

**Comment by karato** [comment]
@dmzaytsev the code made here contains a puzzle `600-f8c0ce2e`/#692, which will be resolved soon

**Comment by karato** [comment]
@ypshenychka please, review this task for compliance with our [quality rules](http://www.teamed.io/qa.html)

**Comment by ypshenychka** [comment]
@karato Quality is good here.

**Comment by karato** [comment]
> @karato Quality is good here.

@ypshenychka thank you :)

**Comment by karato** [comment]
@erimerturk just added 10 mins to @ypshenychka (for QA), payment ID is `59085142`; paid, thanks, added **30 mins** to your account, payment `AP-07G90720MA999790C`, 288 hours and 38 mins was spent; +30 added to your rating, at the moment it is: [+180](http://www.netbout.com/b/35280?open=rating)

**Comment by erimerturk** [comment]
@karato quality is good but i didn't get any bonus for it ? is it normal ? what is the rule of getting bonus ? thank you

**Comment by yegor256** [comment]
@erimerturk at the moment, we don't give any bonuses for good quality. we're planning to start giving them in some time (in a few weeks, I think)

**Comment by karato** [comment]
@dmzaytsev last puzzled solved `692-1caba32d`/#770
